### PR TITLE
Validate export filename

### PR DESCRIPTION
**Pull request by acelaya**
This PR adds basic validation to the export annotations filename.

The validation consists on testing the input value against this regexp `/^[^\n:\\/*?"'<>]{1,255}$/` (ban characters `\n`, `:`, `\`, `/`, `*`, `?`, `"`, `'`, `<` and `>`. Allow filenames between 1 and 255 characters).

If introduced value is invalid, the "export" button is disabled, and the `Input` is marked as erroneous via `hasError` prop.

[screen-recording.webm](https://github.com/hypothesis/client/assets/2719332/fa24de9d-01c7-4e3b-9346-be9df0779e04)

### TODO

- [x] Display error message, or consider styling the input differently (for example, with red border).
- [x] Add/fix tests

### Testing steps

1. Make sure you have the `export_annotations` feature flag enabled here http://localhost:5000/admin/features
2. Go to the export annotations panel.
3. Initially, a valid name should have been suggested, so the "Export" button is enabled.
4. If you introduce an invalid filename (empty or longer than 255 characters, or containing any of these characters: `\n`, `:`, `\`, `/`, `*`, `?`, `"`, `'`, `<`, `>`), the "Export" button should be immediately disabled, and a red ring should appear in the input.

> **Warning**
> Notice there's a small glitch, which makes the red ring not appear right away, but only if the input looses focus, or one more character is input.
>
> I think it's because there's some style conflict with the focus ring, but it could also be a re-rendering issue. I will investigate this later.

### CODE REVIEW

**Review by lyzadanger — CHANGES_REQUESTED** [review]
Hi, @acelaya !

As I mentioned in the sync just now, I think we may be able to find a way to not re-render this component on all input events. What we might do is use state to track whether the filename input is currently valid. We want to re-render if that boolean flips, but the `onInput` callback could validate `inputRef.current.value` and set the is-valid state as needed. Likewise, if you wanted to prevent setting a suggested filename if the user has inputted something, you could compare directly against the current value of the input field (instead of using state).

In summary: I'd suggest not using state for the input field's value and instead reference the element's `value` directly through a ref.

**Comment by acelaya** [comment]
> In summary: I'd suggest not using state for the input field's value and instead reference the element's value directly through a ref.

It probably makes sense. I'll do the changes and update the PR description.

> Likewise, if you wanted to prevent setting a suggested filename if the user has inputted something, you could compare directly against the current value of the input field (instead of using state).

This more or less worked out of the box by combining `defaultValue` (with suggested filename) and `value` (with user input tracked via state).

If the value was not set at all, the `defaultValue` would be displayed, and if the user manually edited the `value`, it would take precedence.

Since now this requires a bit more logic, I think I may leave it out of this PR.

**Comment by lyzadanger** [comment]
I've proposed that we hold on this PR for a day or two and then I can help pick it up and take it to the finish line by:

* Looking at the issue with delayed error styling
* Addressing the behavior on focused-group change
* Adding and styling an error message

I've assigned myself to this PR to reflect this and will move it into draft state.

**Comment by acelaya** [comment]
Thanks Lyza!

File: src/sidebar/components/ShareDialog/ExportAnnotations.tsx [hunk 1]
```diff
@@ -29,6 +31,11 @@ function ExportAnnotations({ annotationsExporter }: ExportAnnotationsProps) {
   const draftCount = store.countDrafts();
 
   const inputRef = useRef<HTMLInputElement | null>(null);
+  const defaultFilename = useMemo(
+    () => suggestedFilename({ groupName: group?.name }),
+    [group],
+  );
+  const [filenameIsValid, setFilenameIsValid] = useState(true);
```
Comment (robertknight) [inline src/sidebar/components/ShareDialog/ExportAnnotations.tsx:38]: The conventional approach here would be to store the input value as state and then derive `filenameIsValid` from that. The state would be initialized to `defaultFilename`.

**Comment by robertknight** [comment]
I had a look into the browser logic related to this. The value of the `download` attribute is treated as a suggestion and the browser already implements the complex and platform-dependent logic of replacing disallowed characters with something else, as well as setting appropriate length limits. In practice all browsers seem to use underscores as a replacement for illegal chars.

**Comment by robertknight** [comment]
To add some additional thoughts here:

- We should definitely ensure that the name is not empty. It might be worth doing this first in a separate PR since it is straightforward.
- Restricting the character set or length is optional, since the browser will modify the download filename for us if needed (usually by replacing unsupported characters with underscores)
- Any restrictions that we do implement need to be added in accessible way, preferably leaning on built-in support that form fields have (see https://github.com/hypothesis/lms/pull/5624 for an example of enforcing length limits his way).
- If we do decide to restrict the character set, we need to clearly indicate which characters are disallowed, as otherwise it will be unclear to users exactly why the form is disabled.

We could also display a warning rather than a validation error if the user has entered characters that we think are likely to be replaced by the browser in the saved file name.

File: src/sidebar/components/ShareDialog/ExportAnnotations.tsx (hunk 1 above)
Comment (acelaya) [reply src/sidebar/components/ShareDialog/ExportAnnotations.tsx:38]: I also feel inclined to make the input a controlled one, as we need a piece of state anyway.

**Comment by acelaya** [comment]
Thanks Rob. I think your suggestions make sense.

I specially like the suggestion about restricting the length via standard HTML attributes, as is will make it more accessible.

I think I will open a separate PR, as this requires a number of changes and it's easier to apply them from scratch.
